pyhms/demes/single_pop_eas/de.py

```diff
@@ -0,0 +1,210 @@
+import numpy as np
+import scipy
+
+from ...core.individual import Individual
+from ...core.population import Population
+from .common import VariationalOperator, apply_bounds
+
+
+def get_randoms(population: Population) -> np.ndarray:
+    choices = np.indices((population.size, population.size))[1]
+
+    # Create a mask to exclude self from selections
+    mask = np.ones(choices.shape, dtype=bool)
+    np.fill_diagonal(mask, 0)
+
+    # Apply mask to choices to exclude self-indices
+    filtered_choices = choices[mask].reshape(population.size, population.size - 1)
+
+    # For each individual, select 3 random indices (without replacing)
+    indices = np.array([np.random.choice(row, size=3, replace=False) for row in filtered_choices])
+
+    # Select the genomes based on these indices
+    return population.genomes[indices]
+
+
+class BinaryMutation(VariationalOperator):
+    def __init__(self, f: float) -> None:
+        self.f = f
+
+    def __call__(self, population: Population) -> Population:
+        randoms = get_randoms(population)
+        donor = randoms[:, 0] + self.f * (randoms[:, 1] - randoms[:, 2])
+        new_genomes = apply_bounds(donor, population.problem.bounds, "reflect")
+        new_fitness = np.where(
+            np.all(new_genomes == population.genomes, axis=1),
+            population.fitnesses,
+            np.nan,
+        )
+        return Population(new_genomes, new_fitness, population.problem)
+
+
+class BinaryMutationWithDither(VariationalOperator):
+    def __call__(self, population: Population) -> Population:
+        randoms = get_randoms(population)
+        scaling = np.random.uniform(0.5, 1, size=population.size)
+        scaling = np.repeat(scaling[:, np.newaxis], len(population.problem.bounds), axis=1)
+        donor = randoms[:, 0] + scaling * (randoms[:, 1] - randoms[:, 2])
+        new_genomes = apply_bounds(donor, population.problem.bounds, "reflect")
+        new_fitness = np.where(
+            np.all(new_genomes == population.genomes, axis=1),
+            population.fitnesses,
+            np.nan,
+        )
+        return Population(new_genomes, new_fitness, population.problem)
+
+
+class CurrentToPBestMutation:
+    def __call__(
+        self,
+        population: Population,
+        archive: Population,
```

Comment:
Current to best mutation is also used for DEs without archive. I see checks for archive being set to None below, so its type here should be updated.